File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#define ABS(x) ((x) > 0 ? (x) : -(x))

using namespace std;

int n, cnt = 0, board[16], visited[16] = {0,};

int isAttack(int y)
{
    for (int i = 1; i < y; i++)
    {
        if (ABS(y - i) == ABS(board[y] - board[i]))
            return 1;
    }

    return 0;
}

void nQueen(int y)
{
    if (y > n)
    {
        ++cnt;
        return ;
    }

    for (int i = 1; i <= n; i++)
    {
        if (!visited[i])
        {
            board[y] = i;
            if (!isAttack(y))
            {
                visited[i] = 1;
                nQueen(y + 1);
                visited[i] = 0;
            }
        }      
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> n;
    nQueen(1);
    cout << cnt;

    return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {

    static int n, cnt = 0;
    static int[] board;
    static boolean[] isQueen;

    public static boolean checkQueen(int idx) {
        if (isQueen[board[idx]]) {
            return false;
        }

        for (int i = 0; i < idx; i++) {
            if (Math.abs(i - idx) == Math.abs(board[i] - board[idx])) {
                return false;
            }
        }

        return true;
    }

    public static void nQueen(int idx) {
        if (idx == n) {
            ++cnt;
            return;
        }

        for (int i = 0; i < n; i++) {
            board[idx] = i;

            if (checkQueen(idx)) {
                isQueen[i] = true;
                nQueen(idx + 1);
                isQueen[i] = false;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        board = new int[n];
        isQueen = new boolean[n];

        nQueen(0);
        System.out.println(cnt);
    }
}